pkg/aggregator/tags/store.go

```diff
@@ -0,0 +1,181 @@
+	return &Store{
+		tagsByKey: map[ckey.TagsKey]*Entry{},
+		enabled:   enabled,
+		telemetry: *newStoreTelemetry(name),
+	}
+}
+
+// Insert returns an Entry that corresponds to the key. If the key is
+// not in the cache, a new entry is stored in the Store with the tags
+// retrieved from the tagsBuffer.
+func (tc *Store) Insert(key ckey.TagsKey, tagsBuffer *tagset.HashingTagsAccumulator) *Entry {
+	if !tc.enabled {
+		return &Entry{
+			tags: tagsBuffer.Copy(),
+			refs: 1,
+		}
+	}
+
+	entry := tc.tagsByKey[key]
+	if entry != nil {
+		entry.refs++
+		tc.telemetry.hits.Inc()
+	} else {
+		entry = &Entry{
+			tags: tagsBuffer.Copy(),
+			refs: 1,
+		}
+		tc.tagsByKey[key] = entry
+		tc.cap++
+		tc.telemetry.miss.Inc()
+	}
+
+	return entry
+}
+
+// Shrink will try to release memory if cache usage drops low enough.
+func (tc *Store) Shrink() {
+	stats := entryStats{}
+	for key, entry := range tc.tagsByKey {
+		if entry.refs == 0 {
+			delete(tc.tagsByKey, key)
+		} else {
+			stats.visit(entry)
+		}
+	}
+
+	if len(tc.tagsByKey) < tc.cap/2 {
+		new := make(map[ckey.TagsKey]*Entry, len(tc.tagsByKey))
+		for k, v := range tc.tagsByKey {
+			new[k] = v
+		}
+		tc.cap = len(new)
+		tc.tagsByKey = new
+	}
+
+	tc.updateTelemetry(&stats)
+}
+
+func (tc *Store) updateTelemetry(s *entryStats) {
+	t := &tc.telemetry
+
+	tlmMaxEntries.Set(float64(tc.cap), t.name)
+	tlmEntries.Set(float64(len(tc.tagsByKey)), t.name)
+
+	for i := 0; i < 3; i++ {
+		tlmTagsetRefsCnt.Set(float64(s.refsFreq[i]), t.name, fmt.Sprintf("%d", i+1))
+	}
+	for i := 3; i < 8; i++ {
+		tlmTagsetRefsCnt.Set(float64(s.refsFreq[i]), t.name, fmt.Sprintf("%d", 1<<(i-1)))
+	}
+
+	tlmTagsetMinTags.Set(float64(s.minSize), t.name)
+	tlmTagsetMaxTags.Set(float64(s.maxSize), t.name)
+	tlmTagsetSumTags.Set(float64(s.sumSize), t.name)
+}
+
+func newCounter(name string, help string, tags ...string) telemetry.Counter {
+	return telemetry.NewCounter("aggregator_tags_store", name,
+		append([]string{"cache_instance_name"}, tags...), help)
+}
+
+func newGauge(name string, help string, tags ...string) telemetry.Gauge {
+	return telemetry.NewGauge("aggregator_tags_store", name,
+		append([]string{"cache_instance_name"}, tags...), help)
+}
+
+var (
+	tlmHits          = newCounter("hits_total", "number of times cache already contained the tags")
+	tlmMiss          = newCounter("miss_total", "number of times cache did not contain the tags")
+	tlmEntries       = newGauge("entries", "number of entries in the tags cache")
+	tlmMaxEntries    = newGauge("max_entries", "maximum number of entries since last shrink")
+	tlmTagsetMinTags = newGauge("tagset_min_tags", "minimum number of tags in a tagset")
+	tlmTagsetMaxTags = newGauge("tagset_max_tags", "maximum number of tags in a tagset")
+	tlmTagsetSumTags = newGauge("tagset_sum_tags", "total number of tags stored in all tagsets by the cache")
+	tlmTagsetRefsCnt = newGauge("tagset_refs_count", "distribution of usage count of tagsets in the cache", "ge")
```

Comment:
What is the meaning of `"ge"`?

Reply:
"greater or equal to", similar to "le" used by prometheus histograms.